### PR TITLE
atomic policy updates

### PR DESCRIPTION
**Pull request by bflood-strike**
Atomic Policy updates and updates to just send slots
## Description

- InitWallet
    - just the slot for signer is specified now when specifying who to make a config approver
    - they are checked against the signers slot in the same request for validity
- WalletConfigPolicyUpdate
    - require all fields (approval counts, time out and config approvers) to be all specified
    - removed the locks
    - just the slot for signer is specified now when specifying who to make a config approver
    - a signers hash which is a hash of all the config approvers keys is also sent and verified in the program
- BalanceAccountCreation
    - require all fields (approval counts, time out and transfer approvers) to be all specified
    - just the slot for signer is specified now when specifying who to make a transfer approver
    - a signers hash which is a hash of all the transfer approvers keys is also sent and verified in the program
- BalanceAccountPolicyUpdate
    - require all fields (approval counts, time out and transfer approvers) to be all specified
    - just the slot for signer is specified now when specifying who to make a transfer approver
    - a signers hash which is a hash of all the transfer approvers keys is also sent and verified in the program 
    - removed the lock
## Motivation and Context

- removed the needs for locks and allow multiple changes in parallel
- allows more approvers to be specified in a message since just the slot is specified

## How Has This Been Tested?

Unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Bug fix breaking (fix that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ x] New feature breaking (change which adds functionality and causes existing functionality to change)

## Checklist:

- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ x] I have added tests to cover my changes.
- [x ] All new and existing tests passed.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -121,6 +121,9 @@ pub enum WalletError {
     /// The operation was initiated by a different version of the program
     #[error("Operation Version Mismatch")]
     OperationVersionMismatch,
+    /// The specified signers hash did not match the programs calculated hash.
+    #[error("Invalid Signers Hash")]
+    InvalidSignersHash,
```
Comment (dgabriele) [inline src/error.rs:126]: This is cool 👍🏼

File: src/handlers/balance_account_policy_update_handler.rs [hunk 2]
```diff
@@ -69,13 +64,12 @@ pub fn finalize(
...
         },
         || -> ProgramResult {
+            let mut wallet = Wallet::unpack(&wallet_account_info.data.borrow_mut())?;
             wallet.update_balance_account_policy(account_guid_hash, update)?;
+            Wallet::pack(wallet, &mut wallet_account_info.data.borrow_mut())?;
             Ok(())
         },
     )?;
 
-    wallet.unlock_balance_account_policy_updates(account_guid_hash)?;
```
Comment (dgabriele) [inline src/handlers/balance_account_policy_update_handler.rs:77]: You are a destroyer of locks

File: src/model/wallet.rs [hunk 3]
```diff
@@ -495,15 +462,16 @@ impl Wallet {
         let (slot_id, mut balance_account) =
             self.get_balance_account_with_slot_id(account_guid_hash)?;
 
-        self.disable_transfer_approvers(&mut balance_account, &update.remove_transfer_approvers)?;
-        self.enable_transfer_approvers(&mut balance_account, &update.add_transfer_approvers)?;
+        balance_account.transfer_approvers.disable_all();
+        self.enable_transfer_approvers_by_slot(&mut balance_account, &update.transfer_approvers)?;
```
Comment (dgabriele) [inline src/model/wallet.rs:466]: nice refactor

File: tests/balance_account_update_tests.rs [hunk 4]
```diff
@@ -630,16 +507,13 @@ async fn invalid_balance_account_policy_updates() {
...
-                        context.approvers[1].pubkey_as_signer(),
-                    )],
+                    approvals_required_for_transfer: 1,
+                    approval_timeout_for_transfer: Duration::from_secs(7200),
+                    transfer_approvers: vec![SlotId::new(0), SlotId::new(1)],
+                    signers_hash,
                 },
             ),
-            Custom(WalletError::InvalidSlot as u32),
+            Custom(WalletError::InvalidSignersHash as u32),
```
Comment (dgabriele) [inline tests/balance_account_update_tests.rs:516]: Nice improvement over the more generic error

File: tests/common/utils.rs [hunk 5]
```diff
@@ -2440,8 +2482,14 @@ pub async fn create_balance_account(
                 approver_keypairs
                     .iter()
                     .enumerate()
-                    .map(|(i, s)| (SlotId::new(i), s.pubkey_as_signer()))
+                    .map(|(i, s)| (SlotId::new(i), s.pubkey_as_signer()).0)
```
Comment (dgabriele) [inline tests/common/utils.rs:2485]: How about `iter().map(|s| SlotId::new(i))`?

**Review by dgabriele — APPROVED** [review]
Looks great, man! Nice little refactors here and there

File: src/model/wallet.rs [hunk 6]
```diff
@@ -713,6 +651,16 @@ impl Wallet {
         }
         Ok(())
     }
+
+    fn hash_signers(&self, signer_slots: &Vec<SlotId<Signer>>) -> Hash {
+        let mut bytes: Vec<u8> = Vec::new();
+        for id in signer_slots {
+            if self.signers[*id] != None {
```
Comment (bholzman) [inline src/model/wallet.rs:658]: I can't put my finger on it, but leaving the bytes out completely when the signer isn't there doesn't feel quite right to me. I don't think there's a valid situation where the signer would be missing but the transaction should succeed, so maybe we should just fail immediately in that case.

**Review by bholzman — APPROVED** [review]
Looks great!

File: tests/common/utils.rs (hunk 5 above)
Comment (bflood-strike) [reply tests/common/utils.rs]: change this slightly per you suggestion

File: src/model/wallet.rs (hunk 6 above)
Comment (bflood-strike) [reply src/model/wallet.rs:656]: refactored this into a validate_signers_hash which will fail if approver not there.

File: src/model/wallet.rs [hunk 7]
```diff
@@ -715,6 +647,26 @@ impl Wallet {
...
     }
+
+    fn validate_signers_hash(
+        &self,
+        signer_slots: &Vec<SlotId<Signer>>,
+        provided_hash: &Hash,
+    ) -> ProgramResult {
+        let mut bytes: Vec<u8> = Vec::new();
+        for id in signer_slots {
+            if self.signers[*id] != None {
```
Comment (bholzman) [inline src/model/wallet.rs:658]: How about:
```
if let Some(signer) = self.signers[*id] {
  bytes.extend_from_slice(signer.key.as_ref());
} else {
  return Err(WalletError::UnknownSigner.into());
}

File: tests/common/utils.rs [hunk 8]
```diff
@@ -2285,6 +2312,15 @@ pub fn random_balance_account_guid_hash() -> BalanceAccountGuidHash {
     BalanceAccountGuidHash::new(&Pubkey::new_unique().to_bytes())
 }
 
+/// Hash array of signer keys
```
Comment (bholzman) [inline tests/common/utils.rs:2315]: nit: technically it's a vector not an array :-)

**Review by bholzman — APPROVED** [review]

File: src/model/wallet.rs (hunk 7 above)
Comment (bflood-strike) [reply src/model/wallet.rs]: changed - thx

File: tests/common/utils.rs (hunk 8 above)
Comment (bflood-strike) [reply tests/common/utils.rs]: fixed

**Review by bholzman — APPROVED** [review]

**Pull request merged**